---
title: "RNAseq analysis on ASO screen"
author: "Jakob Jung"
date: "August 4, 2022"
output:
pdf_document: default
html_document:
df_print: paged
---
Here I perform the downstream analysis of the RNAseq experiment, in which Salmonella 1344 was challenged with different types of ASOs targeting the *acpP* gene. RNAseq was performed to determine transcriptome changes upon exposure to PNA. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = TRUE)
```

# Packages
I import the packages needed for all analysis. They can all be installed from Bioconductor or CRAN if not statet otherwise:
```{r, message=FALSE}
library(edgeR)
library(circlize)
library(dplyr)
library(ggplot2)
library(ggpubr)
library('RUVSeq')
library(RColorBrewer)
library(MetBrewer)
library(EDASeq)
library(ggrepel)
library(viridis)
library(svglite)
library(ComplexHeatmap)
library(tidyverse)
library(grid)
library(circlize)
```

# Data Acquisition  

I generated a tab file containing gene counts after upstream processing. Upstream processing included folowing steps (starting with fastq-files): 
 
 - BBtools for filtering, trimming and mapping 
 - featureCounts for generating a count matrix


Firstly, I import the gene-wise counts:
```{r}
GenewiseCounts <- read.delim(
  "../data/rna_align/counttable.txt",sep = "\t",
  row.names = 1, header = T, comment.char = "#")

dim(GenewiseCounts)
head(GenewiseCounts[,1:6])
```
\hfill\break

I have to change column names, since they include the whole path:
```{r}
gwc <- GenewiseCounts[,5:length(GenewiseCounts[1,])]
pnapat <- "\\.\\.\\.data\\.rna_align\\.[^_]+_(.+)\\.fq\\.gz\\.bam"
colnames (gwc) <- gsub(pnapat,"\\1", colnames(gwc))
#remove samples
gwc <- gwc[,!grepl("PNA22|PNA15|PNA14", colnames(gwc))]

colnames(gwc)
```

# Gene length vs. expression:
make plots visualizing length:
```{r}
# normalized (TPM)
gwc_lex <- gwc[, grepl("(Control)|(Length)", colnames(gwc))]
lex <- data.frame(length=gwc_lex$Length, counts=rowMeans(gwc_lex[,-1]))

gwcnorm_length <- data.frame(sapply(gwc_lex[,-1], function(x) x / (gwc[,1]/1000)))
gwc_tpm <- data.frame(Length = gwc_lex$Length, 
                      sapply(gwcnorm_length, function(x) x * 1e6 / sum(x)), row.names = rownames(gwc_lex))
tpm <- data.frame(length=gwc_tpm$Length, counts=rowMeans(gwc_tpm[,-1]))
```

Make plots:
```{r}
lex %>% ggplot(aes(x=length, y=log10(counts+1))) + geom_point() + 
  scale_x_continuous(limits = c(0,500)) + scale_y_continuous(limits = c(-0.5,5)) + theme_minimal() +
  geom_text_repel(aes( label=ifelse(length<80, rownames(lex), "")), size=2.5, max.overlaps = 15)

tpm %>% ggplot(aes(x=length, y=log10(counts+1))) + geom_point() + 
  scale_x_continuous(limits = c(0,500)) + scale_y_continuous(limits = c(0,5)) + theme_minimal()

```


I also create a factor variable for groups of the sample data:
```{r}
test <- as.factor(gsub("_\\d$", "", colnames(gwc)[-1]))
test
```


# DE analysis 

Now that I have the read count dataframe with sample names, I import them into the edgeR environment:
```{r}
y <- DGEList(gwc[,-1], group = test, genes = gwc[,1,drop=FALSE])
options(digits = 3)
head(y$samples)
```


## Filtering
Now I want to filter out Genes which have very low counts across all libraries. 
I do this by creating a cutoff $$\frac {10} {L} $$
where L is the minimum library size in millions. We delete genes that are below the cutoff in at least 2 libraries:
```{r}
L <- min(y$samples$lib.size) / 1000000
cutoff <- 10/L
keep <- rowSums(cpm(y) > cutoff) >= 5
table(keep)
```
I retain only the unfiltered genes,and delete 519 genes below the threshold:
```{r}
y <- y[keep, , keep.lib.sizes=FALSE]
```



# Design matrix
I create a design matrix for the samples:
```{r}
batches = as.factor(gsub(".*_(\\d)$", "\\1", colnames(y)))

design <- model.matrix(~0+test)
colnames(design) <- c(levels(test))
rownames(design) <- colnames(y$counts)
design[1:5,]
```


# Normalization

I check how the standard TMM normalization of edgeR performs. I start with calculating normalization factors:
```{r, error=FALSE}
y <- calcNormFactors(y)
y <- estimateDisp(y, design, robust = T)
```

And now I create PCA and RLE plots:
```{r}
mycolors <- c("black","coral4","darkgoldenrod3","#1F78B4","cyan4","darkslategrey","deepskyblue4", "#E31A1C","deeppink3",
              "bisque4","#6A3D9A","coral4", "grey")
#mycolors <- met.brewer("Redon", direction = 1)

plotPCA(cpm(y), col=mycolors[test])
plotRLE(cpm(y), outline=FALSE, ylim=c(-1, 1), col=mycolors[test],
        main="RLE", las=2)

logCPM <- cpm(y, log=TRUE, prior.count=2) 
logCPM_no_batch <- removeBatchEffect(logCPM, batch=batches, design=model.matrix(~0+test))

# MDS plot:
plotMDS(logCPM_no_batch, col=mycolors[test], top = 1000)
plotMDS(logCPM, col=mycolors[test], top = 1000)


# PCA:
pca <- prcomp(t(logCPM))
df_pca <- as.data.frame(pca$x)

theme<-theme(panel.background = element_blank(),panel.border=element_rect(fill=NA),
             panel.grid.major = element_blank(),panel.grid.minor = element_blank(),
             strip.background=element_blank(),
             axis.text=element_text(colour="black", size=12),axis.ticks=element_line(colour="black"),
             axis.title=element_text(colour="black", size=13),
             plot.margin=unit(c(1,1,1,1),"line"),legend.position = "none")

percentage <- round(pca$sdev / sum(pca$sdev) * 100, 2)
percentage <- paste( colnames(df_pca), "(", paste( as.character(percentage), "%", ")", sep="") )



df_pca$group <-test
rownames(df)

plotnames <- gsub("PNA_21", "PNA", rownames(y$samples))
plotnames <- gsub("_", "-", plotnames)
plotnames <- gsub("-([12])", "-R\\1", plotnames)
plotnames <- gsub("JVPep112", "KFF", plotnames)

p<-ggplot(df_pca,aes(x=PC1,y=PC2,group=group,label=plotnames, colour=group))
p<-p+geom_point(size=3)+ scale_shape_identity()+
  geom_text_repel(size=4, min.segment.length = 0, seed = 42, box.padding = 0.5, max.overlaps = 20)+
  theme + xlab(percentage[1]) + ylab(percentage[2])+ scale_color_manual(values = mycolors)
p

# save pca as svg
svg("../analysis/PCA.svg")
print(p)
dev.off()
df_pca$group
```

You can see that the TMM was succesful (TMM centers the RLE around 0). The PCA plot shows a slight batch effect, which I address using 


## Differential Expression analysis

Next, I perform differential expression analysis using TMM-normalized dataset:
```{r}
# make a contrast:
con <- makeContrasts(RNA_MOE = RNA_MOE - Control,
                     RNA_F = RNA_F - Control,
                     LNA = LNA - Control,
                     RNA_Ome = RNA_Ome - Control,
                     DNA_PTO = DNA_PTO - Control,
                     PMO = PMO - Control,
                     PNA_21 = PNA_21 - Control,
                    # PNA22 = PNA22 - Control,
                     #PNA15 = PNA15 - Control,
                     #PNA14 = PNA14 - Control,
                     JVPep112 = JVPep112 - Control,
                     KFFN3 = KFFN3 - Control,
                     levels = design)


fit <- glmQLFit(y, design, robust = TRUE)


res <- list(    RNA_MOE = glmQLFTest(fit, contrast = con[,1]),
                 RNA_F = glmQLFTest(fit, contrast = con[,2]), 
                 LNA = glmQLFTest(fit, contrast = con[,3]), 
                 RNA_Ome = glmQLFTest(fit, contrast = con[,4]), 
                 DNA_PTO = glmQLFTest(fit, contrast = con[,5]),
                 PMO = glmQLFTest(fit, contrast = con[,6]),
                 PNA_21 = glmQLFTest(fit, contrast = con[,7]),
               #  PNA22 = glmQLFTest(fit, contrast = con[,8]),
              #   PNA15 = glmQLFTest(fit, contrast = con[,9]),
              #   PNA14 = glmQLFTest(fit, contrast = con[,10]),
            JVPep112 = glmQLFTest(fit, contrast = con[,8]),
            KFFN3 = glmQLFTest(fit, contrast = con[,9]))

```





We now create MD, BCV and QLDisp plots to access qualiy of data:
```{r}
plotMD(y, main = "MD-plot")
abline(h=0, col="red", lty=2, lwd=2)
plotBCV(y)
plotQLDisp(fit)
```
The quality looks decent.


Now I create a function which makes nice volcano-plots and run it on all the results (all PNA-samples are compared to water control for DE):
```{r}
do_volcano <- function(restab, targetgene = NULL, pointsize = 2, x_limit = F,y_limit=F, show_sig = F, alpha=0.05, 
                       minlogfc=1, title = "Volcano", off_target_list = NULL, phopq = NULL) {
  cols = c("targets"="darkorange","0 mm"="cyan", "1 mm"="blue","2 mm"="darkblue", "PhoPQ"="darkred")
  rownames(restab) <- gsub("^([^A-Z].+)" , "italic('\\1')" , rownames(restab))
  g_labels <- gsub("^([^A-Z].+)" , "italic('\\1')" , targetgene)
  g = ggplot(restab) +
  geom_point(
    data = restab,
    aes(x = logFC, y = -log10(FDR)),
    color = "darkgrey",
    cex = pointsize
  ) + theme_bw()+ # change theme to standard black&wite.
  geom_hline(yintercept = -log10(alpha),
             color = "black", linetype = 3) +
  geom_vline(xintercept = c(-minlogfc,minlogfc),
             color = "black", linetype = 3) +
  theme(axis.title.x = element_text(size=20),
        axis.title.y = element_text(size=20),
        axis.text = element_text(size=15, colour = "black"),
        panel.background = element_rect(colour = "black"),
        axis.line = element_line(colour = "black"),
        panel.grid.minor.x = element_blank(),
        panel.grid.minor.y = element_blank(),
        panel.grid.major.x =  element_blank(),#element_line(colour="lightgrey", size=0.3),
        panel.grid.major.y = element_blank(),#element_line(colour="lightgrey", size=0.3),
        plot.title = element_text(hjust = 0.5, size = 23), 
        legend.position = "none")+
  ggtitle(title)+
  xlab(expression("log"[2]*" fold change")) +
  ylab(expression("- log"[10]*" P-value (FDR)"))+
  scale_x_continuous(expand = c(0,0),breaks = seq(-20,20,2), limits = c(-x_limit,x_limit)) +
  scale_y_continuous(expand = c(0, 0),breaks = seq(0,26,2), limits = c(0,y_limit)) 
  
  if (!is.null(phopq)) {
    g <- g + geom_point(
        data = restab[restab$genes %in% phopq,],
        aes(x = logFC, y = -log10(FDR), bg = "PhoPQ"),
        cex = pointsize, pch=21)
  }  
  
  if (is.null(off_target_list)) {
    g <- g + 
      geom_point(
        data = restab[restab$FDR<alpha & restab$logFC < -minlogfc,],
        aes(x = logFC, y = -log10(FDR)),
        color = "steelblue", 
        cex = pointsize) +
      geom_point(
        data = restab[restab$FDR<alpha & restab$logFC > minlogfc,],
        aes(x = logFC, y = -log10(FDR)),
        color = "darkred", 
        cex = pointsize) 
  } else{
    # different colors for different mismatches:
    mm321 <- restab[gsub("^([^A-Z].+)" , "italic('\\1')" , unique(unlist(off_target_list[3]))),]
    mm321 <- mm321[mm321$logFC < -1 & mm321$FDR < 0.001,]
    
    g <- g +
      geom_point(
        data = restab[gsub("^([^A-Z].+)" , "italic('\\1')" , unlist(off_target_list[3])),],
        aes(x = logFC, y = -log10(FDR), color = "2 mm"),
        cex = pointsize) +
      geom_point(
        data = restab[gsub("^([^A-Z].+)" , "italic('\\1')" , unlist(off_target_list[2])),],
        aes(x = logFC, y = -log10(FDR), color = "1 mm"),
        cex = pointsize)  +
    #zero mismatches:
      if (length(off_target_list[[1]]) > 0) {
      geom_point(
        data = restab[gsub("^([^A-Z].+)" , "italic('\\1')" , unlist(off_target_list[1])),],
        aes(x = logFC, y = -log10(FDR), color = "0 mm"),
        cex = pointsize) 
      }
    g_labels <- c(g_labels, gsub("^([^A-Z].+)" , "italic('\\1')" , unlist(off_target_list[1])))
    g_labels <- c(g_labels, rownames(mm321))
    g_labels <- g_labels[!grepl("NA", g_labels)]
  }
  
  # show the sign. genes:
  # show the sigficantest genes:
  if(show_sig){
    range01 <- function(x){(x-min(x))/(max(x)-min(x))}
    top_up <- restab[ which(restab$FDR < alpha & restab$logFC > minlogfc),]
    top_down <- restab[ which(restab$FDR < alpha & restab$logFC < -(minlogfc)),]
    
    if (length(rownames(top_up)) > 0 && (length(rownames(top_up)) > 3)){
    logFC.scaled <- range01(top_up$logFC)
    FDR.scaled <- range01(-log(top_up$FDR))
    summ <- (logFC.scaled + FDR.scaled)
    top_up <- top_up[order(-summ),][1:3,]
    }

    if (length(rownames(top_down))>0 && (length(rownames(top_down))> 3)){
      logFC.scaled <- range01(-top_down$logFC)
      FDR.scaled <- range01(-log(top_down$FDR))
      summ <- (logFC.scaled + FDR.scaled)
      top_down <- top_down[order(-summ),][1:3,]
    }

    top_peaks <- rbind(top_up, top_down)
    top_peaks <- na.omit(top_peaks)


    g_labels <- c(g_labels, rownames(top_peaks))
  }
  
  g <- g + geom_point(
        data = restab[gsub("^([^A-Z].+)" , "italic('\\1')" , targetgene),],
        aes(x = logFC, y = -log10(FDR), color = "targets"),
        cex = pointsize+1)
  
  # add labels:
  g_labels <- unique(g_labels)
  
  g <- g + geom_label_repel(
    data = restab[g_labels,] , aes(x = logFC, y = -log10(FDR), label = gsub("-","_",rownames(restab[g_labels,]))),
    hjust = 0.1,
    size = 4, segment.alpha = 0.5, 
    segment.color = "black", 
    min.segment.length=unit(0, "cm"), parse = T) + scale_fill_manual(values=cols) + scale_color_manual(values=cols)
  g
}
```




```{r}
# I get the links between locus tags and gene names:
pnames <- read.delim("../data/link_lt_gn.tab", header = F)
rownames(pnames) <- pnames$V2

# I also import PhoPQ related genes:
ppq_raw <- read.delim("../data/PHOPQ.tsv", header = T)
ppq <- as.character(ppq_raw$PhoPQ[ppq_raw$PhoPQ != ""])
phopqvolc <- c(pnames[pnames$V1 %in% ppq,]$V2, "PinT", "SL1344_1169", "SL1344_1168")
prefname <- ifelse(phopqvolc %in% pnames$V2 ,pnames[phopqvolc,]$V1, "" )
prefname <- ifelse(isUnique(prefname), prefname, "")
phopqvolc <- ifelse(prefname != "", prefname, phopqvolc)
phopqvolc <- c(phopqvolc, "phoP", "phoQ")
```




Now I adjust p-values (FDR), create volcano plots, histograms for the results (and save volcano plots as pdfs): 
```{r}
list_ot_0 <- list()
list_ot_1 <- list()
list_ot_2 <- list()
l_volcanoplots <- list()

for (resname in names(res)){
  # adjust p-values FDR
  res[[resname]]$table$FDR <- p.adjust(res[[resname]]$table$PValue, method = "fdr")
  restab <- res[[resname]]$table
  restab$locus_tag <- rownames(restab)
  targetgene <-  "acpP"
  
  #add genenames (not locustags)
  prefname <- ifelse(rownames(restab) %in% pnames$V2 ,pnames[rownames(restab),]$V1, "" )
  prefname <- ifelse(isUnique(prefname), prefname, "")
  rownames(restab) <- ifelse(prefname != "", prefname, rownames(restab))
  
  restab$genes <- rownames(restab)
  
  hist(restab$PValue, breaks=100, main=resname)
  

  # make volcanos:
  svg(paste0("../analysis/volcanoplots/",resname, ".svg"))
  print(do_volcano(restab, targetgene = c("acpP","fabF"), title=resname, 
                   x_limit = 9, show_sig = T, 
                   y_limit = 10,
                   alpha=0.001, pointsize = 2))
  dev.off()
  pdf(paste0("../analysis/volcanoplots/",resname, ".pdf"))
  print(do_volcano(restab, targetgene = c("acpP","fabF"), title=resname, 
                 x_limit = 9, show_sig = T, 
                 y_limit = 10,
                 alpha=0.001, pointsize = 3))
  dev.off()
  

  

  #save result_table:
  dataname <- paste("../analysis/diff_exp_rawdata/", resname, ".csv", sep = "")
  write.csv(restab[order(restab$FDR),], dataname)
  #pval distributions:
  hist(restab$PValue, breaks=100, main=resname)
  l_volcanoplots[[resname]] <- do_volcano(restab, targetgene = c("acpP","fabF"), title=resname, 
                 x_limit = 9, show_sig = T, 
                 y_limit = 10,
                 alpha=0.001, pointsize = 3)
}

library(cowplot)
library(gridExtra)
n <- length(l_volcanoplots)
nCol <- 2

pdf("../analysis/volcanoes.pdf", height = 18, width = 16)
grid.arrange(l_volcanoplots$RNA_MOE, l_volcanoplots$RNA_F, l_volcanoplots$LNA, l_volcanoplots$RNA_Ome, l_volcanoplots$DNA_PTO, l_volcanoplots$PMO, l_volcanoplots$PNA_21, l_volcanoplots$JVPep112, l_volcanoplots$KFFN3, ncol = 3)
dev.off()

reg_genes  <- as_tibble(t(sapply(res, function(x) {
  upreg <- nrow(x$table[x$table$FDR<0.001&x$table$logFC>1,])
  downreg <- nrow(x$table[x$table$FDR<0.001&x$table$logFC<(-1),])
  tot <- upreg + downreg
  c(upreg, downreg, tot)
})))

names(reg_genes) <- c("upregulated", "downregulated", "total_regulated")
reg_genes$sample <- names(res)
writexl::write_xlsx(reg_genes, path = "../analysis/analysis_cd_2023_02/nrs_up_down_genes.xlsx", col_names = T)
```
get acpP and fabF levels from all conditions:
```{r}
acpp_fabF_logchange <- as.data.frame(t(sapply(res, function(x) x$table[c("SL1344_1133", "SL1344_1134"),]$logFC)))
acpp_fabF_pvals <- as.data.frame(t(sapply(res, function(x) x$table[c("SL1344_1133", "SL1344_1134"),]$FDR)))

colnames(acpp_fabF_logchange) <- c("acpP", "fabF")
colnames(acpp_fabF_pvals) <- c("acpP", "fabF")

acpp_fabF_logchange$moiety <- rownames(acpp_fabF_logchange)
acpp_fabF_pvals$moiety <- rownames(acpp_fabF_pvals)

acpp_fabF_logchange <- acpp_fabF_logchange %>% pivot_longer(1:2, names_to = "gene", values_to = "log2FC")
acpp_fabF_pvals <- acpp_fabF_pvals %>% pivot_longer(1:2, names_to = "gene", values_to = "pval")

acpp_fabF_logchange$sign <- ifelse(acpp_fabF_pvals$pval < 0.01 & abs(acpp_fabF_logchange$log2FC)>1, "*", "")
acpp_fabF_logchange$moiety <- factor(acpp_fabF_logchange$moiety, levels = unique(acpp_fabF_logchange$moiety))


g_mismatches_acpp_fabf <- acpp_fabF_logchange %>% ggplot(aes(x=moiety, y=-log2FC, fill=gene)) + 
  geom_hline(yintercept=1, color="black", linetype = "dotted") +
  geom_bar(stat='identity', width = 0.8, position = "dodge", color="black") +
  geom_text(stat = "identity" , aes(label=sign), vjust=-0.3,hjust=0.3 , size=5,position = position_dodge(width=.8))+ 
  theme_classic() + ylab(expression(log[2]~fold~gene~depletion)) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1,vjust=1, size = 9),
        axis.text.y = element_text(size=10),
        axis.title = element_text(size=14),
        title = element_text(size=15),
        legend.title = element_text(size=12),
        legend.text = element_text(face="italic", size = 12))+
  scale_y_continuous(limits = c(-0.1,2.5)) + xlab("") + 
  geom_hline(yintercept=0, color="black") + 
   
  scale_fill_manual(values=c('steelblue','lightblue'))
g_mismatches_acpp_fabf

pdf("../analysis/depletion_target_transcripts.pdf", height=5)
print(g_mismatches_acpp_fabf)
dev.off()

svg("../analysis/depletion_target_transcripts.svg", height=5)
print(g_mismatches_acpp_fabf)
dev.off()
```


check for overlapping genes:
```{r}
DEgenes <- sapply(res, function (x) rownames(x$table[x$table$FDR < 0.001 & abs(x$table$logFC),]))
Reduce(intersect, DEgenes)
```


## heatmaps rnaseq:
```{r}
logchange <- data.frame(sapply(res, function(x) x$table$logFC), 
                                row.names = rownames(res$RNA_MOE$table), 
                                genenames = prefname)
pvals <- data.frame(sapply(res, function(x) x$table$FDR), 
                            row.names = rownames(res$RNA_MOE$table),
                            genenames = prefname)

logchange <- logchange[order(pvals$RNA_MOE),]
pvals <- pvals[order(pvals$RNA_MOE),]

#generate data for upset plots:
list_topup_per_cond <- lapply(names(pvals)[!(names(pvals)=="genenames")], function(x) {
  degenes <- pvals[rownames(logchange)[logchange[[x]]<(-1) & pvals[[x]] < 0.001],]
  degenes <- rownames(degenes[order(degenes[[x]]),])
  x <- degenes[1:15]
  x[!is.na(x)]
})

names(list_topup_per_cond) <- names(pvals)[!(names(pvals)=="genenames")]

topdegenes <- c()
for (i in names(logchange[!(names(logchange)=="genenames")])) {
  degenes <- pvals[rownames(logchange)[abs(logchange[[i]])>1 & pvals[[i]] < 0.01],]
  degenes <- rownames(degenes[order(degenes[[i]]),])
  topdegenes <- unique(append(topdegenes, degenes[1:10]))
}
topdegenes <- topdegenes[!is.na(topdegenes)]

#topdegenes <- unique(c(unlist(lapply(res, function(x) c(rownames(topTags(x, n=5)), rownames(topTags(x, n=5, sort.by = "logFC")))))))  


logchange <- logchange[topdegenes,]
pvals <- pvals[topdegenes,]

rownames(logchange) <- ifelse(logchange$genenames == "", rownames(logchange),
                                            logchange$genenames)
rownames(pvals) <- rownames(logchange)

logchange <- logchange[order(logchange$RNA_MOE, decreasing = T),][!colnames(logchange)=="genenames"]
pvals <- pvals[rownames(logchange),][!colnames(pvals)=="genenames"]


diff_exp <- sapply(names(pvals), function(x) {
  tf <- pvals[[x]] < 0.001 & abs(logchange[[x]])>1
  ifelse(tf, "*", " ")
  })

colnames(logchange) <- gsub("PNA_21", "PNA", colnames(logchange))
colnames(logchange) <- gsub("JVPep112", "KFF", colnames(logchange))
colnames(logchange) <- gsub("_", "-", colnames(logchange))
c1 =  circlize::colorRamp2(c(-2, 0, 2), c("steelblue", "white", "darkorange"))
ht4 <- Heatmap(t(logchange), name = "Log2 FC",
               col = c1,
               cluster_rows = F, cluster_columns = F, show_heatmap_legend = F,
               cell_fun = function(j, i, x, y, width, height, fill) {
                 grid.text(sprintf("%.1s", t(diff_exp)[i, j]), x, y)
               }, 
               border = TRUE,
               width = unit(40, "cm"), height = unit(10, "cm"),
               #row_names_max_width = max_text_width(c(0,0),gp = gpar(fontsize = 0)),
               column_names_rot = 45)

lgd2 = Legend(col_fun = c1, title = expression("Log"[2]*" FC"), 
             title_gp = gpar(fontsize = 20), labels_gp = gpar(fontsize = 15), grid_height =  unit(1, "cm"),
             at = c(-2, 0, 2), legend_width = unit(8, "cm"), border = "black",
             labels = c("-2", "  0", "  2"), legend_height = unit(4, "cm"),
             title_position = "topcenter", direction = "horizontal")

svg("../analysis/heatmap_rnaseq_topDEgenes.svg", height = 15, width = 20)
print(ht4)
draw(lgd2, x = unit(25, "cm"), y = unit(29, "cm"), just = c("center", "top"))
dev.off()
```
## Venn diagrams/upset plots:
```{r}
m <- make_comb_mat(list_topup_per_cond[!is.na(list_topup_per_cond)])
UpSet(m[comb_size(m) >= 1])
```



## KEGG:
I perform the KEGG-analysis using the FRY gene set analysis tool from limma. I start with getting KEGGREST:
```{r}
library(KEGGREST)
# get link and list to get kegg info:
link_kegg <- keggLink("pathway", "sey")
list_kegg <- keggList("pathway", "sey")

kegg_pw_ids <- names(list_kegg)

#rename genes, remove ones which arent in our data:
names(link_kegg) <- gsub("sey:(.*)", "\\1", names(link_kegg)) #rename genes as locus tags
```


```{r}
link_kegg <- link_kegg[names(link_kegg) %in% c(rownames(res$RNA_MOE$table))] #remove genes not in data


idx_kegg <- sapply(kegg_pw_ids, function(x){
  x <- unique(names(link_kegg[grepl(x, link_kegg)])) # choose all genes, except duplucates
})
# add phopq pw to kegg
ppq_raw <- read.delim("../data/PHOPQ.tsv", header = T)

for (c in colnames(ppq_raw)) {
  gs <- ppq_raw[[c]][ppq_raw[[c]]!=""]
  gs_lt <- pnames[pnames$V1 %in% gs,]$V2
  idx_kegg[[c]] <- gs_lt[gs_lt %in% rownames(y$counts)]
}
```

```{r}
l <- length(colnames(con))
acpp_kegg_fry <- lapply(1:l, function(x) fry(y,idx_kegg, design, con[,x]))
names(acpp_kegg_fry) <- colnames(con)
```

add KEGG terms:
```{r}
for (fryres in names(acpp_kegg_fry)) {
  acpp_kegg_fry[[fryres]][["TERM"]] <- ifelse(grepl("sey",rownames(acpp_kegg_fry[[fryres]])),
                                              list_kegg[rownames(acpp_kegg_fry[[fryres]])],
                                              rownames(acpp_kegg_fry[[fryres]]))
  acpp_kegg_fry[[fryres]][["TERM"]] <- gsub("(.*) - Salmonella enterica subsp. enterica serovar Typhimurium SL1344",
                                            "\\1", acpp_kegg_fry[[fryres]][["TERM"]])
  write.csv(acpp_kegg_fry[[fryres]], paste("../analysis/pathway_analysis/", fryres, ".csv", sep = ""))
}


acpp_kegg_frysig <- lapply(acpp_kegg_fry, function(x) x[x[["FDR"]]<0.01 & x[["NGenes"]]>10,])
kegg_siggos <- c()


for (i in names(acpp_kegg_frysig)) {
  print(i)
  print(dim(acpp_kegg_frysig[[i]]))
  print(acpp_kegg_frysig[[i]][,c(1,2,4,7)])
  kegg_siggos <- c(kegg_siggos, rownames(acpp_kegg_frysig[[i]][,]))  # can be modified
}

kegg_siggos <- unique(kegg_siggos[!grepl("NA", kegg_siggos)])
```


Create a heatmap-df  for KEGG:
```{r}
idx_kegg_char <- lapply(idx_kegg, as.character)


# I create a dataframe with mean logFC values for each significant GO-term:
hm_kegg <- t(as.data.frame(lapply(idx_kegg_char[kegg_siggos], function(x){
  sapply(names(res), function(y){
    mean(res[[y]]$table[x,]$logFC)
  })
})))


hm_kegg <- as.data.frame(hm_kegg)

rownames(hm_kegg) <- gsub("\\.", "\\:", rownames(hm_kegg))
```
make heatmap:
```{r}
hm_kegg <- hm_kegg[order(hm_kegg[,9], decreasing = T),]

kegg_sizes <- sapply(idx_kegg_char[rownames(hm_kegg)], function(x) length(x))

pvals <- data.frame(sapply(names(acpp_kegg_fry), 
                           function(x) acpp_kegg_fry[[x]][rownames(hm_kegg),"FDR"]),
                    row.names = rownames(hm_kegg))

#select only significant ones:
pvals <-sapply(pvals, function(x) ifelse(x<0.01, x <- "*", x<-"") )

keggpws <- acpp_kegg_fry$RNA_MOE[rownames(hm_kegg),] [["TERM"]]


rownames(hm_kegg) <- ifelse(!is.na(keggpws),keggpws, rownames(hm_kegg) )

colnames(hm_kegg) <- colnames(logchange)
```

plot hm (save as pdf):
```{r}

col_fun = colorRamp2(c(-1,0, 1), c("steelblue", "white", "orange"))


ht_vert <- Heatmap(hm_kegg, cluster_rows = F, cluster_columns = F,
               name = "GO-analysis", col = col_fun,
               show_heatmap_legend = F, 
               row_title_side = "right", row_title_rot = 0,
               border = TRUE, 
               cell_fun = function(j, i, x, y, width, height, fill) {
                 grid.text(sprintf("%.1s", pvals[i, j]), x, y)
               }, 
               column_names_gp = gpar(fontsize = 11),
               row_names_gp = gpar(fontsize = 10),
               row_title = NULL,
               width = unit(12, "cm"), height = unit(18, "cm"),
               
               right_annotation = rowAnnotation(genes = anno_barplot(kegg_sizes)))

ht_vert

lgd = Legend(col_fun = col_fun, title = expression("mean log"[2]*" FC"), #direction = "horizontal",
             title_gp = gpar(fontsize = 17), labels = c("-1", " 0"," 1"), legend_height = unit(6, "cm"),
             grid_width = unit(0.8, "cm"), labels_gp = gpar(fontsize = 13),
             at = c(-1, 0, 1), border = "black",
             title_position = "leftcenter-rot")
draw(lgd)

svg("../analysis/pathway_analysis/hm_KEGG.svg", width = unit(12, "cm"),  height = unit(10, "cm"))
draw(ht_vert)
draw(lgd, x = unit(2, "cm"), y = unit(10, "cm"), just = c("left", "bottom"))
dev.off()

pdf("../analysis/pathway_analysis/hm_KEGG.pdf", width = unit(12, "cm"),  height = unit(10, "cm"))
draw(ht_vert)
draw(lgd, x = unit(2, "cm"), y = unit(10, "cm"), just = c("left", "bottom"))
dev.off()
```


Packages used:
```{r}
sessionInfo()
```

